File: android/liverelay-sdk/src/main/java/app/orizn/liverelay/ConferenceSession.kt
package app.orizn.liverelay

import android.content.Context
import org.webrtc.AudioTrack
import org.webrtc.DataChannel
import org.webrtc.IceCandidate
import org.webrtc.MediaStream
import org.webrtc.MediaStreamTrack
import org.webrtc.PeerConnection
import org.webrtc.RtpReceiver
import org.webrtc.RtpTransceiver
import org.webrtc.VideoTrack
import java.util.concurrent.ConcurrentHashMap

/**
 * Session de conférence N-parties (équivalent Android de `ConferenceSession` dans liverelay.js).
 *
 * Architecture :
 *  - Une peer connection "publish" envoie caméra + micro via POST /sfu/conference (conferenceJoin).
 *  - Une peer connection "subscribe" recvonly PAR participant distant, créée par [subscribeTo]
 *    via POST /sfu/conference/subscribe (conferenceSubscribe + target_peer_id).
 *
 * [start] rejoint la room puis souscrit automatiquement à tous les participants déjà présents.
 * Pour les participants qui rejoignent après, appeler [subscribeTo] (peer_id obtenu via la
 * couche signalisation applicative).
 *
 * Threading : les callbacks de [LiveRelayListener] arrivent sur les threads internes WebRTC —
 * poster vers le main thread côté appelant si nécessaire (mise à jour UI).
 */
class ConferenceSession(
    private val context: Context,
    private val config: LiveRelayConfig,
    private val room: String,
) {

    var listener: LiveRelayListener? = null

    val capture: MediaCapture = MediaCapture(context)

    /** Notre peer_id dans la conférence, disponible après [start]. */
    var peerId: String? = null
        private set

    private val signaling = SignalingClient(config)

    /** Peer connection de publication (sendonly de fait). */
    private var publishPc: PeerConnection? = null

    /** Peer connections de souscription, une par participant distant (peerId -> PC). */
    private val subscribePcs = ConcurrentHashMap<String, PeerConnection>()

    @Volatile
    private var state: SessionState = SessionState.NEW

    @Volatile
    private var stopped = false

    /**
     * Rejoint la conférence : configure l'audio, démarre caméra + micro, publie via
     * conferenceJoin, puis souscrit à chaque participant déjà présent.
     */
    suspend fun start() {
        check(state == SessionState.NEW) { "ConferenceSession cannot be restarted; create a new instance" }
        PeerConnectionProvider.initialize(context)
        setState(SessionState.CONNECTING)

        try {
            AudioConfigurator.configureForCommunication(context)

            capture.startCamera()
            capture.startMicrophone()
            val videoTrack = capture.videoTrack
                ?: throw LiveRelayException("Camera capture failed to produce a video track")
            val audioTrack = capture.audioTrack
                ?: throw LiveRelayException("Microphone capture failed to produce an audio track")

            val iceServers = signaling.fetchIceServers()
            val pc = PeerConnectionProvider.makePeerConnection(iceServers, publishObserver)
            publishPc = pc

            val videoSender = pc.addTrack(videoTrack, listOf(STREAM_ID))
            pc.addTrack(audioTrack, listOf(STREAM_ID))
            LatencyTuning.maintainFramerate(videoSender)

            val offer = PeerConnectionProvider.makeTunedOffer(pc)
            val join = signaling.conferenceJoin(offer, room)
            PeerConnectionProvider.setRemoteAnswer(pc, SdpPayload(sdp = join.sdp, type = join.type))

            peerId = join.peerId
            setState(SessionState.CONNECTED)

            // Souscrit aux participants déjà présents dans la room.
            for (participant in join.participants) {
                if (participant != join.peerId) {
                    subscribeTo(participant)
                }
            }
        } catch (e: Exception) {
            setState(SessionState.FAILED)
            cleanup()
            throw e
        }
    }

    /**
     * Souscrit au flux d'un participant distant via une peer connection recvonly dédiée.
     * Idempotent : no-op si une souscription existe déjà pour ce peer_id.
     * Les tracks distants remontent au [listener] avec ce [peerId].
     */
    suspend fun subscribeTo(peerId: String) {
        if (stopped) throw LiveRelayException("Session is closed")
        if (subscribePcs.containsKey(peerId)) return

        val iceServers = signaling.fetchIceServers()
        val pc = PeerConnectionProvider.makePeerConnection(iceServers, SubscribeObserver(peerId))

        // Réserve la place avant l'échange SDP pour éviter les souscriptions concurrentes.
        val existing = subscribePcs.putIfAbsent(peerId, pc)
        if (existing != null) {
            try {
                pc.close()
            } finally {
                pc.dispose()
            }
            return
        }

        try {
            pc.addTransceiver(
                MediaStreamTrack.MediaType.MEDIA_TYPE_VIDEO,
                RtpTransceiver.RtpTransceiverInit(RtpTransceiver.RtpTransceiverDirection.RECV_ONLY)
            )
            pc.addTransceiver(
                MediaStreamTrack.MediaType.MEDIA_TYPE_AUDIO,
                RtpTransceiver.RtpTransceiverInit(RtpTransceiver.RtpTransceiverDirection.RECV_ONLY)
            )

            val offer = PeerConnectionProvider.makeTunedOffer(pc)
            val answer = signaling.conferenceSubscribe(offer, room, targetPeerId = peerId)
            PeerConnectionProvider.setRemoteAnswer(pc, answer)
            LatencyTuning.tuneReceivers(pc)
        } catch (e: Exception) {
            subscribePcs.remove(peerId, pc)
            try {
                pc.close()
            } finally {
                pc.dispose()
            }
            throw e
        }
    }

    /**
     * Quitte la conférence : ferme la PC de publication et toutes les PCs de souscription,
     * arrête la capture et restaure la config audio. Idempotent.
     */
    fun stop() {
        if (stopped) return
        stopped = true
        cleanup()
        setState(SessionState.CLOSED)
    }

    private fun cleanup() {
        publishPc?.let { pc ->
            try {
                pc.close()
            } finally {
                pc.dispose()
            }
        }
        publishPc = null

        for ((_, pc) in subscribePcs) {
            try {
                pc.close()
            } finally {
                pc.dispose()
            }
        }
        subscribePcs.clear()

        capture.stop()
        AudioConfigurator.reset(context)
    }

    private fun setState(newState: SessionState) {
        if (state == newState || state == SessionState.CLOSED) return
        state = newState
        listener?.onStateChanged(newState)
    }

    /** Observer de la PC de publication : pilote l'état global de la session. */
    private val publishObserver = object : BaseObserver() {
        override fun onIceConnectionChange(newState: PeerConnection.IceConnectionState?) {
            if (stopped) return
            when (newState) {
                PeerConnection.IceConnectionState.CONNECTED,
                PeerConnection.IceConnectionState.COMPLETED -> setState(SessionState.CONNECTED)
                PeerConnection.IceConnectionState.DISCONNECTED -> setState(SessionState.DISCONNECTED)
                PeerConnection.IceConnectionState.FAILED -> setState(SessionState.FAILED)
                else -> Unit
            }
        }
    }

    /** Observer d'une PC de souscription : remonte les tracks distants avec le peer_id associé. */
    private inner class SubscribeObserver(private val remotePeerId: String) : BaseObserver() {
        override fun onAddTrack(receiver: RtpReceiver?, mediaStreams: Array<out MediaStream>?) {
            if (stopped) return
            when (val track = receiver?.track()) {
                is VideoTrack -> listener?.onRemoteVideoTrack(track, remotePeerId)
                is AudioTrack -> listener?.onRemoteAudioTrack(track, remotePeerId)
                else -> Unit
            }
        }
    }

    /** Observer no-op de base — le SFU n'utilise pas le trickle ICE ni les data channels. */
    private abstract class BaseObserver : PeerConnection.Observer {
        override fun onSignalingChange(newState: PeerConnection.SignalingState?) {}
        override fun onIceConnectionChange(newState: PeerConnection.IceConnectionState?) {}
        override fun onIceConnectionReceivingChange(receiving: Boolean) {}
        override fun onIceGatheringChange(newState: PeerConnection.IceGatheringState?) {}
        override fun onIceCandidate(candidate: IceCandidate?) {}
        override fun onIceCandidatesRemoved(candidates: Array<out IceCandidate>?) {}
        override fun onAddStream(stream: MediaStream?) {}
        override fun onRemoveStream(stream: MediaStream?) {}
        override fun onDataChannel(dataChannel: DataChannel?) {}
        override fun onRenegotiationNeeded() {}
        override fun onAddTrack(receiver: RtpReceiver?, mediaStreams: Array<out MediaStream>?) {}
    }

    private companion object {
        const val STREAM_ID = "liverelay-cam"
    }
}

File: ios/Sources/LiveRelaySDK/ConferenceSession.swift
import Foundation
import WebRTC

/// Session de conférence N-parties LiveRelay.
///
/// Flux (miroir de `ConferenceSession` / `subscribeToParticipant` de liverelay.js) :
/// - `start()` : audio session → capture caméra+micro → PC principale (publish)
///   → offre tunée → `POST /sfu/conference` → answer + `peer_id` + `participants`
///   → une PC recvonly par participant déjà présent via `subscribeTo(peerId:)`.
/// - `subscribeTo(peerId:)` : PC dédiée recvonly (video+audio, H.264 préféré)
///   → `POST /sfu/conference/subscribe` avec `target_peer_id`.
/// - Les tracks distantes remontent via `LiveRelaySessionDelegate` avec le
///   `peerId` du participant correspondant.
public final class ConferenceSession {

    public weak var delegate: LiveRelaySessionDelegate?

    /// Capture locale (caméra + micro) publiée dans la conférence.
    public var capture: MediaCapture { mediaCapture }

    /// Notre peer_id dans la conférence (disponible après `start()`).
    public private(set) var peerId: String?

    // MARK: - Privé

    private let config: LiveRelayConfig
    private let room: String
    private let signaling: SignalingClient
    private let provider = PeerConnectionProvider.shared
    private let mediaCapture: MediaCapture

    private let lock = NSLock()
    private var publishPC: RTCPeerConnection?
    private var subscribePCs: [String: RTCPeerConnection] = [:]
    /// `RTCPeerConnection.delegate` est weak : on retient les observers ici.
    private var observers: [PeerConnectionObserver] = []
    private var state: SessionState = .new

    public init(config: LiveRelayConfig, room: String) {
        self.config = config
        self.room = room
        self.signaling = SignalingClient(config: config)
        self.mediaCapture = MediaCapture(factory: PeerConnectionProvider.shared.factory)
    }

    deinit {
        stop()
    }

    // MARK: - Start

    /// Rejoint la conférence : publie caméra+micro puis s'abonne à chaque
    /// participant déjà présent.
    public func start() async throws {
        setState(.connecting)
        do {
            // 1. Audio session + capture locale.
            try AudioSessionConfigurator.configureForVoiceChat()
            try await mediaCapture.startCamera()
            mediaCapture.startMicrophone()

            // 2. Peer connection de publication.
            let iceServers = try await signaling.fetchIceServers()
            let observer = PeerConnectionObserver()
            let pc = provider.makePeerConnection(iceServers: iceServers, delegate: observer)

            observer.onIceConnectionStateChange = { [weak self] iceState in
                self?.handlePublishIceState(iceState)
            }
            // La PC principale peut aussi recevoir des tracks (participants
            // présents au join, stream id "lr-XXXXXXXX" côté serveur).
            observer.onTrack = { [weak self] receiver, streams in
                let remotePeerId = streams.first?.streamId
                self?.deliverTrack(from: receiver, peerId: remotePeerId)
            }

            lock.lock()
            publishPC = pc
            observers.append(observer)
            lock.unlock()

            // 3. Tracks locales + tuning encodeur (avant createOffer).
            if let videoTrack = mediaCapture.videoTrack,
               let sender = pc.add(videoTrack, streamIds: ["liverelay-cam"]) {
                LatencyTuning.maintainFramerate(on: sender)
            }
            if let audioTrack = mediaCapture.audioTrack {
                _ = pc.add(audioTrack, streamIds: ["liverelay-cam"])
            }
            for transceiver in pc.transceivers where transceiver.mediaType == .video {
                LatencyTuning.preferH264(on: transceiver)
            }

            // 4. Offre tunée → conference join → answer.
            let offer = try await provider.makeTunedOffer(for: pc)
            let response = try await signaling.conferenceJoin(offer: offer, room: room)
            try await provider.setRemoteAnswer(
                SdpPayload(sdp: response.sdp, type: response.type),
                on: pc
            )
            LatencyTuning.tuneReceivers(of: pc)

            lock.lock()
            peerId = response.peerId
            lock.unlock()

            setState(.connected)

            // 5. Une PC recvonly par participant déjà présent.
            //    Best-effort : l'échec d'un abonnement ne casse pas la conférence.
            for participant in response.participants where participant != response.peerId {
                do {
                    try await subscribeTo(peerId: participant)
                } catch {
                    #if DEBUG
                    print("[LiveRelay] conference: subscribe to \(participant) failed: \(error)")
                    #endif
                }
            }
        } catch {
            setState(.failed)
            teardown()
            throw error
        }
    }

    // MARK: - Subscribe

    /// S'abonne à un participant (présent au join, ou arrivé après nous).
    public func subscribeTo(peerId targetPeerId: String) async throws {
        lock.lock()
        let alreadySubscribed = subscribePCs[targetPeerId] != nil
        let closed = state == .closed
        lock.unlock()
        guard !alreadySubscribed else { return }
        guard !closed else { throw LiveRelayError.notConnected }

        let iceServers = try await signaling.fetchIceServers()
        let observer = PeerConnectionObserver()
        let pc = provider.makePeerConnection(iceServers: iceServers, delegate: observer)

        observer.onTrack = { [weak self] receiver, _ in
            self?.deliverTrack(from: receiver, peerId: targetPeerId)
        }

        lock.lock()
        subscribePCs[targetPeerId] = pc
        observers.append(observer)
        lock.unlock()

        do {
            // Transceivers recvonly video + audio (H.264 préféré avant l'offre).
            let recvOnly = RTCRtpTransceiverInit()
            recvOnly.direction = .recvOnly
            if let videoTransceiver = pc.addTransceiver(of: .video, init: recvOnly) {
                LatencyTuning.preferH264(on: videoTransceiver)
            }
            _ = pc.addTransceiver(of: .audio, init: recvOnly)

            let offer = try await provider.makeTunedOffer(for: pc)
            let answer = try await signaling.conferenceSubscribe(
                offer: offer,
                room: room,
                targetPeerId: targetPeerId
            )
            try await provider.setRemoteAnswer(answer, on: pc)
            LatencyTuning.tuneReceivers(of: pc)
        } catch {
            lock.lock()
            subscribePCs[targetPeerId] = nil
            lock.unlock()
            pc.close()
            throw error
        }
    }

    // MARK: - Stop

    /// Quitte la conférence : ferme la PC de publication, toutes les PCs
    /// d'abonnement et arrête la capture locale.
    public func stop() {
        lock.lock()
        let alreadyClosed = state == .closed
        lock.unlock()
        guard !alreadyClosed else { return }

        teardown()
        setState(.closed)
    }

    // MARK: - Internals

    private func teardown() {
        lock.lock()
        let publish = publishPC
        let subs = subscribePCs
        publishPC = nil
        subscribePCs.removeAll()
        observers.removeAll()
        lock.unlock()

        for (_, pc) in subs { pc.close() }
        publish?.close()
        mediaCapture.stop()
    }

    private func setState(_ newState: SessionState) {
        lock.lock()
        guard state != newState else {
            lock.unlock()
            return
        }
        state = newState
        lock.unlock()

        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            self.delegate?.session(self, didChangeState: newState)
        }
    }

    private func handlePublishIceState(_ iceState: RTCIceConnectionState) {
        lock.lock()
        let closed = state == .closed
        lock.unlock()
        guard !closed else { return }

        switch iceState {
        case .connected, .completed:
            setState(.connected)
        case .disconnected:
            setState(.disconnected)
        case .failed:
            setState(.failed)
        default:
            break
        }
    }

    private func deliverTrack(from receiver: RTCRtpReceiver, peerId remotePeerId: String?) {
        guard let track = receiver.track else { return }
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            if let video = track as? RTCVideoTrack {
                self.delegate?.session(self, didReceiveVideoTrack: video, peerId: remotePeerId)
            } else if let audio = track as? RTCAudioTrack {
                self.delegate?.session(self, didReceiveAudioTrack: audio, peerId: remotePeerId)
            }
        }
    }
}

// MARK: - PeerConnectionObserver

/// Adaptateur `RTCPeerConnectionDelegate` → closures.
/// `RTCPeerConnection` ne retient pas son delegate : la session le garde en vie.
private final class PeerConnectionObserver: NSObject, RTCPeerConnectionDelegate {

    var onTrack: ((RTCRtpReceiver, [RTCMediaStream]) -> Void)?
    var onIceConnectionStateChange: ((RTCIceConnectionState) -> Void)?

    func peerConnection(_ peerConnection: RTCPeerConnection,
                        didAdd rtpReceiver: RTCRtpReceiver,
                        streams mediaStreams: [RTCMediaStream]) {
        onTrack?(rtpReceiver, mediaStreams)
    }

    func peerConnection(_ peerConnection: RTCPeerConnection,
                        didChange newState: RTCIceConnectionState) {
        onIceConnectionStateChange?(newState)
    }

    // MARK: Méthodes requises restantes (no-op)

    func peerConnection(_ peerConnection: RTCPeerConnection, didChange stateChanged: RTCSignalingState) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didAdd stream: RTCMediaStream) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove stream: RTCMediaStream) {}
    func peerConnectionShouldNegotiate(_ peerConnection: RTCPeerConnection) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceGatheringState) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didGenerate candidate: RTCIceCandidate) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove candidates: [RTCIceCandidate]) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didOpen dataChannel: RTCDataChannel) {}
}
